openhands/runtime/modal/runtime.py

```diff
@@ -0,0 +1,347 @@
+        self.log_stream_thread = threading.Thread(target=self.stream_logs)
+        self.log_stream_thread.daemon = True
+        self.log_stream_thread.start()
+
+
+class ModalRuntime(EventStreamRuntime):
+    """This runtime will subscribe the event stream.
+
+    When receive an event, it will send the event to runtime-client which run inside the Modal sandbox environment.
+
+    Args:
+        config (AppConfig): The application configuration.
+        event_stream (EventStream): The event stream to subscribe to.
+        sid (str, optional): The session ID. Defaults to 'default'.
+        plugins (list[PluginRequirement] | None, optional): List of plugin requirements. Defaults to None.
+        env_vars (dict[str, str] | None, optional): Environment variables to set. Defaults to None.
+    """
+
+    container_name_prefix = 'openhands-sandbox-'
+
+    def __init__(
+        self,
+        config: AppConfig,
+        event_stream: EventStream,
+        sid: str = 'default',
+        plugins: list[PluginRequirement] | None = None,
+        env_vars: dict[str, str] | None = None,
+        status_message_callback: Callable | None = None,
+    ):
+        assert config.modal_api_token_id, 'Modal API token id is required'
+        assert config.modal_api_token_secret, 'Modal API token secret is required'
+
+        self.config = config
+
+        self.modal_client = modal.Client.from_credentials(
+            config.modal_api_token_id, config.modal_api_token_secret
+        )
+        self.app = modal.App.lookup(
+            'openhands', create_if_missing=True, client=self.modal_client
+        )
+
+        # workspace_base cannot be used because we can't bind mount into a sandbox.
+        if self.config.workspace_base is not None:
+            logger.warning(
+                'Setting workspace_base is not supported in the modal runtime.'
+            )
+
+        # This value is arbitrary as it's private to the container
+        self.container_port = 3000
+
+        self.session = requests.Session()
+        self.instance_id = (
+            sid + '_' + str(uuid.uuid4()) if sid is not None else str(uuid.uuid4())
+        )
+        self.status_message_callback = status_message_callback
+
+        self.send_status_message('STATUS$STARTING_RUNTIME')
+        self.base_container_image_id = self.config.sandbox.base_container_image
+        self.runtime_container_image_id = self.config.sandbox.runtime_container_image
+        self.action_semaphore = threading.Semaphore(1)  # Ensure one action at a time
+
+        logger.info(f'ModalRuntime `{self.instance_id}`')
+
+        # Buffer for container logs
+        self.log_buffer: LogBuffer | None = None
+
+        if self.config.sandbox.runtime_extra_deps:
+            logger.debug(
+                f'Installing extra user-provided dependencies in the runtime image: {self.config.sandbox.runtime_extra_deps}'
+            )
+
+        self.image = self._get_image_definition(
```

Comment:
Will this image get cached on modal, or do we have to re-build it everytime when a new runtime is launched?

Reply:
This will be cached. We have caching at each layer in the image build. For each successive `run_command`, `copy_local_file`, etc, we cache the result.